builtin/logical/database/dbs/connectionproducer.go

```diff
@@ -0,0 +1,254 @@
+package dbs
+
+import (
+	"crypto/tls"
+	"database/sql"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/hashicorp/vault/helper/certutil"
+	"github.com/hashicorp/vault/helper/tlsutil"
+	"github.com/mitchellh/mapstructure"
+)
+
+type ConnectionProducer interface {
+	Connection() (interface{}, error)
+	Close()
+	// TODO: Should we make this immutable instead?
+	Reset(*DatabaseConfig) error
+}
+
+// sqlConnectionProducer impliments ConnectionProducer and provides a generic producer for most sql databases
+type sqlConnectionDetails struct {
+	ConnectionURL string `json:"connection_url" structs:"connection_url" mapstructure:"connection_url"`
+}
+
+type sqlConnectionProducer struct {
+	config *DatabaseConfig
+	// TODO: Should we merge these two structures make it immutable?
+	connDetails *sqlConnectionDetails
+
+	db *sql.DB
+	sync.Mutex
+}
+
+func (cp *sqlConnectionProducer) Connection() (interface{}, error) {
+	// Grab the write lock
+	cp.Lock()
+	defer cp.Unlock()
+
+	// If we already have a DB, we got it!
+	if cp.db != nil {
+		if err := cp.db.Ping(); err == nil {
+			return cp.db, nil
+		}
+		// If the ping was unsuccessful, close it and ignore errors as we'll be
+		// reestablishing anyways
+		cp.db.Close()
+	}
+
+	// Otherwise, attempt to make connection
+	conn := cp.connDetails.ConnectionURL
+
+	// Ensure timezone is set to UTC for all the conenctions
+	if strings.HasPrefix(conn, "postgres://") || strings.HasPrefix(conn, "postgresql://") {
+		if strings.Contains(conn, "?") {
+			conn += "&timezone=utc"
+		} else {
+			conn += "?timezone=utc"
+		}
+	} else {
+		conn += " timezone=utc"
+	}
+
+	var err error
+	cp.db, err = sql.Open(cp.config.DatabaseType, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set some connection pool settings. We don't need much of this,
+	// since the request rate shouldn't be high.
+	cp.db.SetMaxOpenConns(cp.config.MaxOpenConnections)
+	cp.db.SetMaxIdleConns(cp.config.MaxIdleConnections)
+
+	return cp.db, nil
+}
+
+func (cp *sqlConnectionProducer) Close() {
+	// Grab the write lock
+	cp.Lock()
+	defer cp.Unlock()
+
+	if cp.db != nil {
+		cp.db.Close()
+	}
+
+	cp.db = nil
+}
+
+func (cp *sqlConnectionProducer) Reset(config *DatabaseConfig) error {
```

Comment:
So in the new RFC it describes a new path `/database/reset/foo` I was thinking of moving this functionality into that path handler and just making the database object immutable. So the path handler would close out the old one and build a new instance of the database object by reading the configs from disk. What do you think of that idea?